src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala

```diff
@@ -0,0 +1,346 @@
+    fields.foldLeft(hasher) { case (h, field) =>
+      //TODO: is there need to check for nulls in v?
+      val v = r.get(field.name())
+      field.schema().getType match {
+        case Type.RECORD => hashAvroField(r, field.schema().getFields.asScala, h)
+        case Type.ENUM => h.putString(v.asInstanceOf[Enum[_]].name(), utf8Charset)
+        case Type.ARRAY =>
+          throw new UnsupportedOperationException("Array is not supported as sampling key!")
+        case Type.MAP =>
+          throw new UnsupportedOperationException("Map is not supported as sampling key!")
+        case Type.UNION =>
+          throw new UnsupportedOperationException("Union is not supported as sampling key!")
+        case Type.FIXED =>
+          throw new UnsupportedOperationException("Fixed is not supported as sampling key!")
+        case Type.STRING => h.putString(v.asInstanceOf[CharSequence], utf8Charset)
+        case Type.BYTES => h.putBytes(v.asInstanceOf[Array[Byte]])
+        case Type.INT => h.putInt(v.asInstanceOf[Int])
+        case Type.LONG => h.putLong(v.asInstanceOf[Long])
+        case Type.FLOAT => h.putFloat(v.asInstanceOf[Float])
+        case Type.DOUBLE => h.putDouble(v.asInstanceOf[Double])
+        case Type.BOOLEAN => h.putBoolean(v.asInstanceOf[Boolean])
+        case Type.NULL =>
+          throw new UnsupportedOperationException("Null is not supported as sampling key!")
+      }
+    }
+  }
+  // scalastyle:on cyclomatic.complexity
+
+  private def diceElement[T](e: T,
+                             hash: HashCode,
+                             sampleSize: Int,
+                             module: Int): TraversableOnce[T] = {
+    //TODO: for now leave it up to jit/compiler to optimize
+    if (math.abs(hash.padToLong()) % module < sampleSize) {
+      Iterable(e)
+    } else {
+      Iterable.empty
+    }
+  }
+
+  private def parseAsBigQueryTable(tblRef: String): Option[TableReference] = {
+    Try(BigQueryIO.parseTableSpec(tblRef)).toOption
+  }
+
+  private def parseAsURI(uri: String): Option[URI] = {
+    Try(new URI(uri)).toOption
+  }
+
+  // scalastyle:off cyclomatic.complexity
+  private def hashTableRow(r: TableRow,
+                           tblSchema: List[TableFieldSchema],
+                           fields: Seq[String],
+                           hasher: Hasher): Hasher = {
+    //TODO: do we need to check for nulls
+    fields.foldLeft(hasher) { case (h, f) =>
+      val fieldParts = f.split('.')
+      val fieldOpt = tblSchema.find(_.getName == fieldParts.head)
+      if (fieldOpt.isEmpty) {
+        log.warn(s"Can't find field ${fieldParts.head} in table schema")
+        hasher
+      } else {
+        val v = r.get(fieldParts.head)
+        val field = fieldOpt.get
+        if (field.getMode == "REPEATED") {
+          //TODO: support repeated fields
+          throw new UnsupportedOperationException(
+            "Repeated field is not supported as sampling key!")
+        }
+        field.getType match {
```

Comment:
Double check the BQ data type spec. IIRC one of `INTEGER` or `FLOAT` is not really `Long` or `Double` but a `String` because of JSON. So `asInstanceOf` will fail.
Also `v` might be `null` if the field is NULLABLE.